src/LazyArray.h

```diff
@@ -0,0 +1,127 @@
+
+#pragma once
+
+/** A dynamic array that defers allocation to the first modifying access.
+Const references from the array before allocation will all be to the same default constructed value.
+It is therefore important that default constructed values are indistinguishable from each other. */
+template <typename T>
+class cLazyArray
+{
+	static_assert(!(std::is_reference<T>::value || std::is_array<T>::value),
+		"cLazyArray<T>: T must be a value type");
+	static_assert(std::is_default_constructible<T>::value,
+		"cLazyArray<T>: T must be default constructible");
+public:
+	using value_type = T;
+	using pointer = T *;
+	using const_pointer = const T *;
+	using reference = T &;
+	using const_reference = const T &;
+	using size_type = int;
+	using iterator = pointer;
+	using const_iterator = const_pointer;
+	
+	cLazyArray(size_type a_Size) NOEXCEPT:
+		m_Size{ a_Size }
+	{
+		ASSERT(a_Size > 0);
+	}
+
+	cLazyArray(const cLazyArray & a_Other):
+		m_Size{ a_Other.m_Size }
+	{
+		if (a_Other.IsStorageAllocated())
+		{
+			std::copy(a_Other.begin(), a_Other.end(), begin());
+		}
+	}
+
+	cLazyArray(cLazyArray && a_Other) NOEXCEPT:
+		m_Array{ std::move(a_Other.m_Array) },
+		m_Size{ a_Other.m_Size }
+	{
+	}
+
+	cLazyArray & operator = (const cLazyArray & a_Other)
+	{
+		cLazyArray(a_Other).swap(*this);
+		return *this;
+	}
+
+	cLazyArray & operator = (cLazyArray && a_Other) NOEXCEPT
+	{
+		m_Array = std::move(a_Other.m_Array);
+		m_Size = a_Other.m_Size;
+		return *this;
+	}
+
+	T & operator [] (size_type a_Idx)
+	{
+		ASSERT((0 <= a_Idx) && (a_Idx < m_Size));
+		return data()[a_Idx];
+	}
+
+	const T & operator [] (size_type a_Idx) const
+	{
+		return GetAt(a_Idx);
+	}
+
+	// STL style interface
+
+	const T * cbegin() const { return data(); }
+	T *        begin()       { return data(); }
+	const T *  begin() const { return cbegin(); }
+
+	const T * cend() const { return data() + m_Size; }
+	T *        end()       { return data() + m_Size; }
+	const T *  end() const { return cend(); }
```

Comment:
You've defined typedefs/using for the iterator types. They should be used here. Then I don't need to change it on 7 places.

Reply:
Fixed